Video: common video API skeleton and muted autoplay for <amp-video>

aghassemi opened the pull request:
Partial for: https://github.com/ampproject/amphtml/issues/4154 and https://github.com/ampproject/amphtml/issues/4113

First PR for the common video API which also brings muted autoplay to `<amp-video>` in supported platforms.

Demo: http://goo.gl/c5fCEC  - Visit in desktop or Chrome Dev (53) in Android

Please note:
- I am writing tests, but wanted to get the review started. 
- UX may change, I meeting with UX folks to make/validate some decisions around autoplay behaviour
- I do not believe this needs an experiment flag. Only affects `<amp-video>` with autoplay attribute which would not have worked in most platforms anyway.

cramforce commented on src/custom-element.js:
```diff
@@ -832,6 +826,20 @@ function createBaseAmpElementProto(win) {
+   * @param {!Object=} opt_data Event data.
+   * @final @this {!Element}
+   */
+  ElementProto.dispatchCustomEventForTest = function(name, opt_data) {
```
Make it `ForTesting`. Our coding convention automatically kills those.

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.video.showControls();
+        this.video.unmute();
```
Did we decide to not support a `muted` attribute? If not, then we should only unmute if muted is not present.

aghassemi replied:
deprecated `muted` attribute.
We may not end up unmuting here depending on what comes out of UX review tomorrow however.

cramforce replied:
Any result? I'm 99% certain we want to unmute if `muted` is not present.

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+      }
+
+      // Calculate what percentage of the video is in viewport
+      const change = this.video.element.getIntersectionChangeEntry();
```
This should be documented properly, but I believe it is fine to call this without a `measure` block, because it exclusively hits methods that are cached and only written to when things change.

dvoytenko replied:
Correct. Measure block is not needed. These are always cached.

cramforce replied:
Thinking about it, it is probably still good to have it, if we need the
mutate block, as that makes sure the measurements are the latest at the
time of mutate.

On Wed, Aug 31, 2016 at 2:29 PM, Dima Voytenko <notifications@github.com>
wrote:

> In src/service/video-manager-impl.js
> <https://github.com/ampproject/amphtml/pull/4786#discussion_r77077194>:
>
> > +   * Called by all possible events that might change the visibility of the video
> > +   * such as scrolling or {@link ../video-interface.VideoEvents#VISIBILITY}.
> > +   * @package
> > +   */
> > +  updateVisibility() {
> > +    const wasVisible = this.isVisible_;
> > +
> > +    // Measure if video is now in viewport and what percentage of it is visible
> > +    const measure = () => {
> > +      if (!this.video.isInViewport()) {
> > +        this.isVisible_ = false;
> > +        return;
> > +      }
> > +
> > +      // Calculate what percentage of the video is in viewport
> > +      const change = this.video.element.getIntersectionChangeEntry();
>
> Correct. Measure block is not needed. These are always cached.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/4786/files/5e55b6681c220ac6c11372aa688dc175547184a0#r77077194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFeT5vAmvPhIXXyRQfFHg8H3D1aBzmaks5qlfJDgaJpZM4JxI8M>
> .
>

aghassemi replied:
ack, left as is.

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+  // TODO(aghassemi): Test other combinations and add support.
+  // TODO(aghassemi): Is there a way to detect that autoplay has been disabled
+  // in the user preferences of the browser?
+  return false;
```
How about assuming `true` and change the conditionals above to return `false`? That would be more robust towards other browsers.

aghassemi replied:
For now I rather white list. If we apply autoplay behaviour in a browser that does not actually support it, we end up with a bad UX (for instance by hiding the controls for no good reason).

I also removed iOS support until I do proper testing.

cramforce replied:
I'd really rather turn on iOS right away. Did you find an iOS 10 device? Simulator is probably fine.

cramforce commented:
This is awesome!!!

I tried the demo and one thing that didn't work for me was that clicking unmuted the video and that (I think this is not implemented) a video that has been interacted with (click on autoplay) no longer pauses when scrolling out of view. I suppose the latter might only make sense when we start keeping these videos in viewport.

jridgewell commented on src/video-interface.js:
```diff
@@ -0,0 +1,165 @@
+   *
+   * @event amp:video:canplay
+   */
+  CAN_PLAY: 'amp:video:built',
```
This doesn't match the documented value.

dvoytenko replied:
Can play is the only video-specific event here. But I believe we just should use the existing video events and just polyfill them for all other players: https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Media_events

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -45,11 +52,20 @@ export function installVideo(win) {
             'No "poster" attribute has been provided for amp-video.');
       }
 
+      videoManagerForDoc(this.win.document).register(this);
```
Should it be unregistered as well?

dvoytenko replied:
Also, what concerns me about `register` is the risk conditions. We obviously don't have it with `amp-video` and not with `autoplay`, but others elements/operations might have it. For instance, "hero" videos, as determined by the video manager, would have to wait for an element to be upgraded/built which might cause some risk conditions if video manager needs to make a decision which video to play. Similarly, if we need to implement a feature "only autoplay one video at a time", this creates risk condition if two videos are in the viewport.

mkhatib replied:
What would be the alternative to this. Querying DOM elements for specific tags (`amp-video` for now) when the video manager is installed and registering them then?

jridgewell commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+    this.win_ = ampdoc.win;
+
+    /** @private {?Array<{!../video-interface.VideoInterface}>} */
+    this.entries_;
```
This is a noop in js, you'll need to set a value.

dvoytenko replied:
`= null;`

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
       return !!this.video_.play;
     }
+
+    /**
```
No jsdoc here. Instead you can add this info to each method as `/** @override See VideoInterface */` or such.

aghassemi replied:
done

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
+     */
+
+    /**
+     * @return {boolean}
```
`@override` on all of these

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
+     * @return {boolean}
+     */
+    hasAutoplay() {
+      return this.element.hasAttribute('autoplay');
```
Any way to standardize `autoplay` attribute across all video elements?

aghassemi replied:
yep, now `autoplay` and `controls` are expected attributed and no longer part of the JS interface (but documented same way the events are in the `video-interface.js`)

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
+     */
+    play(unusedIsAutoplay) {
+      this.video_.play();
+      return listenOncePromise(this.video_, 'play');
```
If it's already playing, will this event never received/promise never yield?

aghassemi replied:
as discussed, removed the promise part of the API.

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
+     */
+    pause() {
+      this.video_.pause();
+      return listenOncePromise(this.video_, 'pause');
```
Ditto, if it's not currently playing. E.g. because it finished or already paused?

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -98,6 +121,80 @@ export function installVideo(win) {
+     * @return {boolean}
+     */
+    hasControls() {
+      return this.element.hasAttribute('controls');
```
Ditto: can we standardize?

dvoytenko commented on src/custom-element.js:
```diff
@@ -511,7 +511,7 @@ function createBaseAmpElementProto(win) {
     if (this.everAttached) {
       this.implementation_.firstAttachedCallback();
-      this.dispatchCustomEvent('amp:attached');
+      this.dispatchCustomEventForTest('amp:attached');
```
/cc @zhouyx who already did in another PR :)

dvoytenko commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+    this.entries_ = this.entries_ || [];
+    const entry = new VideoEntry(this.win_, video);
+
+    if (entry.needsVisibilityObserver()) {
```
Most likely all other features will need visibility observer anyway. Fine to keep until then, of course.

dvoytenko commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
```
Should this be done after built or layout?

dvoytenko commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+
+    // TODO(aghassemi, #4780): Create a new IntersectionObserver service.
+    if (!this.scrollListenerInstalled_) {
+      const scrollListener = () => {
```
We shouldn't do that. Instead we should add API to custom element itself to be notified of the intersection changes.

dvoytenko commented on src/service/viewer-impl.js:
```diff
@@ -152,7 +152,7 @@ export class Viewer {
     this.messagingOrigin_ = null;
 
-    /** @private {!Array<!{eventType: string, data: *}>} */
+    /** @private {!Array<{!eventType: string, data: *}>} */
```
The previous version was correct. `!` should be in front of `{`

dvoytenko commented on src/video-interface.js:
```diff
@@ -0,0 +1,165 @@
+
+
+/**
+ * >>>>>>Work-In-Progress<<<<<<
```
Remove

dvoytenko commented on src/video-interface.js:
```diff
@@ -0,0 +1,165 @@
+ * @interface
+ * @extends {./base-element.BaseElement}
+ */
+export class VideoInterface {
```
I'd definitely also expect `getState()` of some sort here. E.g. "playing", "paused", "finished" or such.

aghassemi replied:
yep, we would need that soon. Not needed yet though.

dvoytenko commented on src/video-interface.js:
```diff
@@ -0,0 +1,165 @@
+   *
+   * @return {boolean}
+   */
+  hasAutoplay() {}
```
As discussed offline, some of these should be just standardized on the element level via attributes.

dvoytenko commented on src/video-interface.js:
```diff
@@ -0,0 +1,165 @@
+   *
+   * @event amp:video:built
+   */
+  BUILT: 'amp:video:built',
```
Hmm. We should just rely on custom element's events/APIs for `build` or `layout`. It's totally reasonable to just do: `CustomElement.whenBuilt()` API and such.

dvoytenko commented on builtins/amp-video.js:
```diff
@@ -45,11 +52,20 @@ export function installVideo(win) {
       this.applyFillContent(this.video_, true);
       this.element.appendChild(this.video_);
+
+      this.element.dispatchCustomEvent(VideoEvents.BUILT);
```
But if we do use `videoManager.register(this)` here, then why do we need `BUILT` event? Registration is essentially equivalent to being built.

aghassemi replied:
as discussed offline, removed built event and registering is expected to happens after component builds.

mkhatib commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,310 @@
+    dev().assert(video);
+
+    // TODO(aghassemi): Remove this later. For now, VideoManager only matters
+    // for autoplay videos so no point in registering arbitrary videos yet.
```
Not sure if planned, but would be awesome if video manager also paused other playing videos when the user explicitly plays a new video on the page.

aghassemi replied:
yep, it is planned. we are doing bunch of UX review meetings to come up with more use-cases that we can support.

aghassemi commented:
Alright, PTAL. 
Tests and iOS support to come in another CL as I like to get this in for Chrome 53 sooner if possible.

cramforce commented on builtins/amp-video.js:
```diff
@@ -98,6 +115,63 @@ export function installVideo(win) {
+
+    /**
+     * @override
+     * {@see ../src/video-interface.VideoInterface}
```
We don't typically add these. Visual noise.

cramforce commented on builtins/amp-video.js:
```diff
@@ -98,6 +115,63 @@ export function installVideo(win) {
+    /**
+     * @override
+     * {@see ../src/video-interface.VideoInterface}
+     * @return {boolean}
```
No need to say unless you specialize a type union.

cramforce commented on builtins/amp-video.md:
```diff
@@ -82,9 +83,10 @@ Similar to the `video` tag `controls` attribute - if present, the browser offers
 If present, will automatically loop the video back to the start upon reaching the end.
 
-**muted**
+**muted (deprecated)**
```
Don't fully understand why. Probably asked before.

aghassemi replied:
Bunch of reasons:

- `muted` definitely make sense when combined with autoplay but we handle that automatically when `autoplay` is present. Essentially we are no longer supporting `unmuted autoplay` even on desktop as of this PR (`unmuted autoplay` is bad UX regardless of desktop or mobile) so when there is `autoplay`, there is no longer a need for `muted` attribute.

- building consistent expectation on what happens when user taps a video (regardless of autoplay). It should either always unmute or never unmute (Can't really see a use-case where a publisher can justifiably break that expectation for the user. User can always mute/change volume themselves)

- most importantly: in iOS, there is no unmute button! if a video is marked as `muted`, user can not unmute it (crazy). Do we want publishers to prevent a user from ever unmuting a video?

I am not super opposed to keeping this attribute, but I rather try to do this for the sake of better UX and if there is a backlash or reasonable use-cases for it, then bring it back again.

/cc @rudygalfi for thoughts.

rudygalfi replied:
`unmuted autoplay` isn't inherently bad UX. It depends on the use case. For instance, if the user lands on a page where the who point it watching a video (like YouTube) playing unmuted by default makes sense. Keeping this, but positioning it as a desktop only feature might be the right way to go.

Regarding iOS, one idea that comes to mind is similar to the bottom video in your demo where it's just for design. If the video is specified as muted, this could be an indicator that it should always just stay muted. By note exposing an unmute control, this might be the stance that iOS is taking. (Although in my use case, it wouldn't make sense to expose any controls.)

aghassemi replied:
Here is scenario that breaks down if we keep it and honour it for `autoplay`:

* Let's say we want **muted autoplay** on **desktop and mobile**, but users should be able to **unmute the video as well**. (Essentially the most common use-case)

What's the markup for this scenario?
a) `<amp-video muted autoplay>`
b) `<amp-video autoplay>`

Neither will do what we want because:
a) will give us unmuted autoplay, but user won't be able to unmute in iOS
b) will give us unmuted autoplay on mobile but not on desktop.

If we want to support unmuted autoplay on desktop, we should provide a new attribute maybe `autoplay-with-sound` 

If we keep `muted` but not honour it for autoplay, then we are back to `muted` attribute being mostly useless, because all it can do is decide if tapping a video (autoplay or not) plays the video with sound or not (with the big caveat that iOS users won't be able to unmute it because there is no mute/unmute control in iOS)

cramforce replied:
My main point is that `muted` would be for when you actually don't want to play the sound. Like a video that has sound, but you want to loop it and put it in the background.

aghassemi replied:
yeah and that case is covered by `autoplay` and lack of `controls`. `<amp-video autoplay loop>` would exactly do what you are describing.

What we are missing now is ability to autoplay with sound which would only work in desktop. I suggest not worrying about that case yet, if we need to have it, we can add support either using `autoplay=withsound` or `autoplay-with-sound`. More interesting (and more challenging) would be trying to carry user's intention of watching a video all the way from the platform in a way that browser would allow us to call `play()` on unmuted videos to get unmuted autoplay in mobile.

cramforce replied:
OK, I don't completely follow the logic, since having a standard attribute
just do what it does seems easier to document than anything else, but don't
want to bikeshed it. My recommendation would be to say: muted, OK, this
video will not have sound.

Let me know when the rest the other change are made.

On Mon, Sep 12, 2016 at 4:37 PM, Ali Ghassemi <notifications@github.com>
wrote:

> In builtins/amp-video.md
> <https://github.com/ampproject/amphtml/pull/4786#discussion_r78474467>:
>
> > @@ -82,9 +83,10 @@ Similar to the `video` tag `controls` attribute - if present, the browser offers
> >
> >  If present, will automatically loop the video back to the start upon reaching the end.
> >
> > -**muted**
> > +**muted (deprecated)**
>
> yeah and that case is covered by autoplay and lack of controls. <amp-video
> autoplay loop> would exactly do what you are describing.
>
> What we are missing now is ability to autoplay with sound which would only
> work in desktop. I suggest not worrying about that case yet, if we need to
> have it, we can add support either using autoplay=withsound or
> autoplay-with-sound. More interesting (and more challenging) would be
> trying to carry user's intention of watching a video all the way from the
> platform in a way that browser would allow us to call play() on unmuted
> videos to get unmuted autoplay in mobile.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/4786/files/c2a77b4669e07783c5adc2dc14b5835b7bb91884#r78474467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFeT1Qvi4M4WDwV0Hppe5eWCwCcQIkTks5qpeIjgaJpZM4JxI8M>
> .
>

aghassemi replied:
The main issue is that UA's interpret `muted` attribute differently. Here is an example, what do you think this will do:
`<video muted autoplay controls>`?
Safari in iOS 10 interprets this as the video has no audio tracks and does not allow the user to  unmute this. (just try https://jsfiddle.net/q65768nw/ in iOS 10, there is no unmute button, volume controls doesn't do anything. Chrome gives users an unmute UI button with its default state dictated by the `muted` attribute.)

ok crap, so how is the author gonna fix that? They will remove the `muted` attribute. Now what's gonna happen?

oh crap, nothing autoplays on mobile platforms anymore. What to do now?...

At least until Safari in iOS 10 gives users a UI control to unmute a muted auto-playing video, what this attribute does natively is both confusing and can create a situation where users can not unmute a video! So we can't simply document and follow what native `unmute` attribute does.

dvoytenko replied:
Was there consensus on this point? Should we involve anyone on any browser's side and get some ideas where things are most likely to be in the future?

aghassemi replied:
How about starting on conversation on whatwg/html to clarify what `muted` attribute implies for the UA. The way iOS is interpreting it is very restrictive. The spec is not very clear either: "muted - Whether to mute the media resource by default". It does not say if UA should still allow the user to unmute a muted video or not.

dvoytenko replied:
Good idea. Please start and link to one of our issues.

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,311 @@
+    if (this.isVisible_) {
+      this.video.play(/*autoplay*/ true);
+    } else {
+      this.video.pause();
```
I've been thinking about this and I don't think we should pause a video that has been interacted with. Even if we don't have the "sticky video" feature yet I don't want to stop listening to the recipe just because I scroll down read more ingredients.

mkhatib replied:
I added this in the docs, but `pause` might need a parameter `isAutopause` similar to `isAutoplay` for `play` method. And can be used here to indicate that pause should only happen if the video hasn't been interacted with in the implementation of `pause` method.

aghassemi commented:
working through some ridiculous* safari bugs, please hold on review until I fix them
*If you hide controls, play the video and show the controls again, instead of "pause" button Safari iOS shows "play" button despite the fact that video is actually playing...

mkhatib commented:
One more thing to keep in mind for the autoplaying behavior is to [address accessibility issues](https://www.abilitynet.org.uk/news-blogs/why-autoplay-accessibility-issue) that comes with it. Browsers provide a setting to disable autoplay for these users but we need to make sure that the autoplay logic we have fails gracefully when that setting is on.

aghassemi commented:
@mkhatib Interesting, never thought of a11y issues with autoplay. I have a todo to detect whether autoplay is enabled or not aside from simple UA sniffing, unfortunately there is no API to ask the browser, but there are some tricks. Currently the experience doesn't completely break down if we wrongly assume autoplay (only controls will be hidden) but we should do better for sure.

aghassemi commented:
@dvoytenko @cramforce Alright, code-wise this is ready to go so PTAL. Tested with real iOS 10 and fixed some issues also added the logic that if user has interacted with an autoplaying video, then we no longer do anything special with that video.

cramforce commented on src/base-element.js:
```diff
@@ -482,6 +482,22 @@ export class BaseElement {
   }
 
   /**
+   * Utility method that bubbles events from the given element
```
Should this be called `forwardEvents`. I think it is not clear enough what this does. Am I getting it correctly:

- It listens for non-bubbling events
- and transports them to bubbling custom events of the name name

?

aghassemi replied:
renamed and comments updated.

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,318 @@
+
+/**
+ * Detects whether the platform even supports autoplay videos.
+ * @private
```
No need for `@private`, but add `@param` and `@return`

aghassemi replied:
done

cramforce commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,318 @@
+  }
+
+  // TODO(aghassemi): Test other combinations and add support.
+  // TODO(aghassemi): Is there a way to detect that autoplay has been disabled
```
In preparation for that you may consider to return `!Promise<boolean>` instead `boolean`, so that a slightly async feature test can be done.

aghassemi replied:
the change will ripple through bunch sync functions. Will do it when add the proper autoplay detection feature.

cramforce commented on src/base-element.js:
```diff
@@ -482,6 +482,22 @@ export class BaseElement {
+   * @param  {!Element} element
+   * @protected @final
+   */
+  bubbleEvents(events, element) {
```
Add a test

aghassemi replied:
done

cramforce commented:
A few more things, but basically there. Thanks!

jridgewell commented on src/base-element.js:
```diff
@@ -482,6 +482,22 @@ export class BaseElement {
   /**
+   * Utility method that bubbles events from the given element
+   * to this element.
+   * @param  {!Array<string>} events
```
Can we make this take a `string|!Array<string>`? It'll make single events cleaner to bubble.

aghassemi replied:
done (also for propagateAttributes)

jridgewell commented on builtins/amp-video.js:
```diff
@@ -101,6 +124,80 @@ export function installVideo(win) {
+      }
+
+      this.propagateAttributes(['autoplay'], this.video_);
+      listenOnce(this.video_, VideoEvents.CANPLAY, () => {
```
I realize this is a workaround to get a `CANPLAY` event, but what does it get us after that?

aghassemi replied:
nothing, we clean up after we get the CANPLAY. It is a `once` listener and then we remove the `autoplay` attribute we propagated, so we should be back to a clean state after canplay fires.

jridgewell replied:
So what's the point the hack?

aghassemi replied:
sorry I didn't see your replies until now. 
This a purely a hack that's is required for iOS. 
iOS (wrongly) does not fire 'canplay' event for muted inline videos (I have filed a bug with them). We rely on that event. iOS however does fire that event if `autoplay` attribute is set. Note that we do not set `autoplay` attribute on video elements since we manage autoplay behaviour ourselves (reason being browser's UX implementation of when to start playing/pausing is not inline with that we want). So what this  hack does is 1- sets `autoplay` to ensure iOS fires the 'canplay` event, then 2- removes the `autoplay` attribute after we got what we wanted.

jridgewell replied:
What I'm asking is why do we care about the `canplay` event?

aghassemi replied:
sorry, misunderstood the discussion here. so VideoManager needs the `canplay` event to know if it can call `play()` or not. For instance if a autoplay video becomes visible but it is still not ready to play, video manager has to wait until it becomes ready before calling `play()` on it. `canplay` is the event that ensures `play()` method can be called. Calling `play()` before `canplay` event is fired, will through exception.

jridgewell commented on src/service/video-manager-impl.js:
```diff
@@ -0,0 +1,318 @@
+      return;
+    }
+
+    this.entries_ = this.entries_ || [];
```
Why not just initialize `#entries` to an array to begin with?

aghassemi replied:
tiny allocation optimization since, at least for now, we only register autoplay videos, so many pages won't have anything to register hence the lazy initialization.

jridgewell replied:
But it's an empty array instance? You're gonna waste more bytes in size than you would in memory.

aghassemi replied:
true, initializing it at declaration is more readable too. ok I will change it in one of the next video related PRs.

cramforce commented:
LGTM

aghassemi commented:
Just fixing type errors and running gulp size

jridgewell commented on src/video-interface.js:
```diff
@@ -76,8 +75,23 @@ export class VideoInterface {
+   *
+   * @return {!AmpElement}
+   */
+  get element() {}
```
This is _very_ slow in browsers.

dvoytenko replied:
I think this is just an interface definition, not the actual implementation.

cramforce replied:
No idea if this actually works, but should be harmless.

aghassemi replied:
@jridgewell  yeah it's just interface def for type checking. It gets DC-ed. But good to know that getters are very slow!
@cramforce it does! Without this, compiler errored whenever I did `videoInterfaceInstance.element` saying element is not part of the interface definition.

jridgewell replied:
👍

dvoytenko commented:
LGTM with one question
